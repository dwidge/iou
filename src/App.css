* {
box-sizing: border-box;
}

nav {
display: flex;
justify-content: space-between;
overflow: scroll;
}

.page{
overflow: scroll;
}

.link,.link-active {
color: inherit;
text-decoration: inherit;
border: solid 1px orange;
padding: 1em;
display: block;
}

.link-active {
background: orange;
}

button {
padding: 1em;
margin: 0em;
border: solid 1px orange;
}

textarea {
width: 100%;
height: 10em;
}
