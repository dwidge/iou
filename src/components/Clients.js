import React from 'react'
import PropTypes from 'prop-types'
import { Table, ColumnText } from '@dwidge/table-react'
import { uuid } from '@dwidge/lib'

const Clients = ({ stClients, newId = uuid }) =>
	(<>
		<Table name='Clients' schema={{
			ref: ColumnText('Ref'),
			name: ColumnText('Name'),
			phone: ColumnText('Phone'),
			patterns: ColumnText('Patterns'),
		}} newRow={() => ({ id: uuid(), ref: newId(), name: 'client', phone: '555', patterns: '' })} rows={stClients} />
	</>)

Clients.propTypes = {
	stClients: PropTypes.array.isRequired,
	newId: PropTypes.func,
}

export default Clients
