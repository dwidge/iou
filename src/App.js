import React, { useState, useEffect } from 'react'
import { Storage } from '@dwidge/lib-react'

import Clients from './components/Clients'
import Invoices from './components/Invoices'
import Receipts from './components/Receipts'
import Bank from './components/Bank'
import Statements from './components/Statements'
import Settings from './components/Settings'
import Navbar from './components/Navbar'
import { newRef, useKeyVal } from './components/lib'

// import './App.css'
const { useStorage } = Storage(useState, useEffect)

const defaultSettings = {
	id: 1,
	clnpre: 'C',
	clnnext: 1,
	stmpre: 'S',
	stmnext: 1,
	recpre: 'R',
	recnext: 1,
	invpre: 'I',
	invnext: 1,
}

const App = () => {
	const stClients = useStorage('iou_clients', [])
	const stInvoices = useStorage('iou_invoices', [])
	const stReceipts = useStorage('iou_receipts', [])
	const stSettings = useKeyVal(useStorage('iou_settings', {}), defaultSettings)

	return (
		<Navbar brand='IoU' pages={{
			'/clients': { name: 'Clients', element: (<Clients stClients={stClients} newId={newRef(stSettings)('clnpre', 'clnnext')} />) },
			'/invoices': { name: 'Invoices', element: (<Invoices stInvoices={stInvoices} aClients={stClients[0]} stSettings={stSettings} />) },
			'/receipts': { name: 'Receipts', element: (<Receipts stReceipts={stReceipts} aClients={stClients[0]} stInvoices={stInvoices} stSettings={stSettings} />) },
			'/bank': { name: 'Bank', element: (<Bank stClients={stClients} stInvoices={stInvoices} stSettings={stSettings} stReceipts={stReceipts} newClientId={newRef(stSettings)('clnpre', 'clnnext')}/>) },
			'/statements': { name: 'Statements', element: (<Statements aClients={stClients[0]} aInvoices={stInvoices[0]} aReceipts={stReceipts[0]} stSettings={stSettings} />) },
			'/settings': { name: 'Settings', element: (<Settings stSettings={stSettings} />) },
		}}/>
	)
}

export default App
