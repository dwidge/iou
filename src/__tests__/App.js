import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import * as J from '@dwidge/lib-react'

import React, { useState } from 'react'
import {
	MemoryRouter,
} from 'react-router-dom'
import App from '../App'
import Clients from '../components/Clients'
const tid = id => screen.getByTestId(id)
const text = J.tools(userEvent, screen, jest).text
const clear = J.clear(userEvent, screen)
const type = J.type(userEvent, screen)
const click = J.click(userEvent, screen)
const input = async (id, text) => {
	await clear(id)
	await type(id, text)
}

jest.mock('@dwidge/lib', () => ({
	...jest.requireActual('@dwidge/lib'),
	uuid: ((value = 10) => () => String(++value))(),
}))
afterEach(() => {
	jest.restoreAllMocks()
})

describe('Clients', () => {
	const client1 = { id: 1, name: 'clientA', phone: '1235551234' }

	test('renders page', () => {
		render(<MemoryRouter initialEntries={['/clients']}>
			<App />
		</MemoryRouter>)

		expect(tid('tableClients')).toMatchSnapshot()
	})

	it('enters into list', async () => {
		const Page = () => (<Clients stClients={useState([])}/>)
		render(<Page/>)
		await click('buttonAdd')
		await input('inputName', 'clientB')
		await input('inputPhone', '0005551112')
		await click('buttonSave')
		expect(tid('tableClients')).toMatchSnapshot()
	})

	it('enters another into list', async () => {
		const Page = () => (<Clients stClients={useState([client1])}/>)
		render(<Page/>)
		expect(tid('tableClients')).toMatchSnapshot()
		await click('buttonAdd')
		await input('inputName', 'clientC')
		await input('inputPhone', '0005551113')
		await click('buttonSave')
		expect(tid('tableClients')).toMatchSnapshot()
	})

	it('confirms & clears list', async () => {
		const Page = () => (<Clients stClients={useState([client1])}/>)
		render(<Page/>)
		expect(tid('tableClients')).toMatchSnapshot()
		expect(text('buttonClear')).toEqual('Clear')
		await click('buttonClear')
		expect(text('buttonClear')).toEqual('Confirm')
		expect(tid('tableClients')).toMatchSnapshot()
		await click('buttonClear')
		expect(text('buttonClear')).toEqual('Clear')
		expect(tid('tableClients')).toMatchSnapshot()
	})
})

describe('Invoices', () => {
	test('renders page', () => {
		render(<MemoryRouter initialEntries={['/invoices']}>
			<App />
		</MemoryRouter>)

		expect(tid('tableInvoices')).toMatchSnapshot()
	})
})

describe('Receipts', () => {
	test('renders page', () => {
		render(<MemoryRouter initialEntries={['/receipts']}>
			<App />
		</MemoryRouter>)

		expect(tid('tableReceipts')).toMatchSnapshot()
	})
})

describe('Statements', () => {
	test('renders page', () => {
		render(<MemoryRouter initialEntries={['/statements']}>
			<App />
		</MemoryRouter>)

		expect(tid('tableStatements')).toMatchSnapshot()
	})
})

describe('Settings', () => {
	test('renders page', () => {
		render(<MemoryRouter initialEntries={['/settings']}>
			<App />
		</MemoryRouter>)

		expect(tid('tableSettings')).toMatchSnapshot()
	})
})
