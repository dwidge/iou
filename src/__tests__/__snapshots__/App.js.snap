// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clients confirms & clears list 1`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <column-text />
          </td>
          <td>
            <column-text>
              clientA
            </column-text>
          </td>
          <td>
            <column-text>
              1235551234
            </column-text>
          </td>
          <td>
            <column-text />
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonEdit1"
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonDel1"
              type="button"
            >
              Del
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Clients confirms & clears list 2`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <column-text />
          </td>
          <td>
            <column-text>
              clientA
            </column-text>
          </td>
          <td>
            <column-text>
              1235551234
            </column-text>
          </td>
          <td>
            <column-text />
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonEdit1"
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonDel1"
              type="button"
            >
              Del
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Confirm
  </button>
</div>
`;

exports[`Clients confirms & clears list 3`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Clients enters another into list 1`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <column-text />
          </td>
          <td>
            <column-text>
              clientA
            </column-text>
          </td>
          <td>
            <column-text>
              1235551234
            </column-text>
          </td>
          <td>
            <column-text />
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonEdit1"
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonDel1"
              type="button"
            >
              Del
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Clients enters another into list 2`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <column-text />
          </td>
          <td>
            <column-text>
              clientA
            </column-text>
          </td>
          <td>
            <column-text>
              1235551234
            </column-text>
          </td>
          <td>
            <column-text />
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonEdit1"
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonDel1"
              type="button"
            >
              Del
            </button>
          </td>
        </tr>
        <tr>
          <td>
            <column-text>
              14
            </column-text>
          </td>
          <td>
            <column-text>
              clientC
            </column-text>
          </td>
          <td>
            <column-text>
              0005551113
            </column-text>
          </td>
          <td>
            <column-text />
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonEdit13"
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonDel13"
              type="button"
            >
              Del
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Clients enters into list 1`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <column-text>
              12
            </column-text>
          </td>
          <td>
            <column-text>
              clientB
            </column-text>
          </td>
          <td>
            <column-text>
              0005551112
            </column-text>
          </td>
          <td>
            <column-text />
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonEdit11"
              type="button"
            >
              Edit
            </button>
          </td>
          <td>
            <button
              class="btn btn-primary"
              data-testid="buttonDel11"
              type="button"
            >
              Del
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Clients renders page 1`] = `
<div
  data-testid="tableClients"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Name
          </th>
          <th>
            Phone
          </th>
          <th>
            Patterns
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Invoices renders page 1`] = `
<div
  data-testid="tableInvoices"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            InvoiceRef
          </th>
          <th>
            ClientRef
          </th>
          <th>
            Date
          </th>
          <th>
            Summary
          </th>
          <th>
            Total
          </th>
          <th>
            Send
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Receipts renders page 1`] = `
<div
  data-testid="tableReceipts"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            Ref
          </th>
          <th>
            Date
          </th>
          <th>
            Client
          </th>
          <th>
            Total
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
  <button
    class="btn btn-primary"
    data-testid="buttonAdd"
    type="button"
  >
    Add
  </button>
  <button
    class="btn btn-primary"
    data-testid="buttonClear"
    type="button"
  >
    Clear
  </button>
</div>
`;

exports[`Settings renders page 1`] = `
<div
  data-testid="tableSettings"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <tbody>
        <tr>
          <td
            colspan="7"
          >
            <div>
              <div>
                <b>
                  Client Prefix
                </b>
              </div>
              <div>
                <column-text>
                  C
                </column-text>
              </div>
            </div>
            <div>
              <div>
                <b>
                  Client Next
                </b>
              </div>
              <div>
                <column-text>
                  1
                </column-text>
              </div>
            </div>
            <div>
              <div>
                <b>
                  Invoice Prefix
                </b>
              </div>
              <div>
                <column-text>
                  I
                </column-text>
              </div>
            </div>
            <div>
              <div>
                <b>
                  Invoice Next
                </b>
              </div>
              <div>
                <column-text>
                  1
                </column-text>
              </div>
            </div>
            <div>
              <div>
                <b>
                  Receipt Prefix
                </b>
              </div>
              <div>
                <column-text>
                  R
                </column-text>
              </div>
            </div>
            <div>
              <div>
                <b>
                  Receipt Next
                </b>
              </div>
              <div>
                <column-text>
                  1
                </column-text>
              </div>
            </div>
            <div>
              <div>
                <b />
              </div>
              <div>
                <button
                  class="btn btn-primary"
                  data-testid="buttonEdit1"
                  type="button"
                >
                  Edit
                </button>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Statements renders page 1`] = `
<div
  data-testid="tableStatements"
>
  <div
    style="overflow: auto; max-height: 80vh;"
  >
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            ClientRef
          </th>
          <th>
            BalanceDue
          </th>
          <th>
            Send
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
